src/meta/src/rpc/server.rs

```diff
@@ -171,315 +153,244 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
+                Ok(_) => tracing::info!("Shutting down follower services"),
+                Err(_) => tracing::error!("Follower service receiver dropped"),
+            }
+            // Wait until follower service is down
+            handle.await.unwrap();
+        }
 
-    let (barrier_scheduler, scheduled_barriers) =
-        BarrierScheduler::new_pair(hummock_manager.clone(), env.opts.checkpoint_frequency);
+        let elect_coord = ElectionCoordination {
+            election_handle,
+            election_shutdown,
+        };
 
-    let source_manager = Arc::new(
-        SourceManager::new(
-            barrier_scheduler.clone(),
-            catalog_manager.clone(),
-            fragment_manager.clone(),
+        start_leader_srv(
+            meta_store,
+            address_info,
+            max_heartbeat_interval,
+            opts,
+            leader_rx,
+            elect_coord,
+            svc_shutdown_rx,
         )
         .await
-        .unwrap(),
-    );
-
-    let barrier_manager = Arc::new(GlobalBarrierManager::new(
-        scheduled_barriers,
-        env.clone(),
-        cluster_manager.clone(),
-        catalog_manager.clone(),
-        fragment_manager.clone(),
-        hummock_manager.clone(),
-        source_manager.clone(),
-        meta_metrics.clone(),
-    ));
-
-    {
-        let source_manager = source_manager.clone();
-        tokio::spawn(async move {
-            source_manager.run().await.unwrap();
-        });
-    }
+        .expect("Unable to start leader services");
+    });
 
-    let stream_manager = Arc::new(
-        GlobalStreamManager::new(
-            env.clone(),
-            fragment_manager.clone(),
-            barrier_scheduler.clone(),
-            cluster_manager.clone(),
-            source_manager.clone(),
-            hummock_manager.clone(),
-        )
-        .unwrap(),
-    );
+    Ok((join_handle, svc_shutdown_tx))
+}
 
-    hummock_manager
-        .purge_stale(
-            &fragment_manager
-                .list_table_fragments()
+// TODO: repeat test 100 times. Start test with different sleep intervals and so on
+// Print the sleep intervals
+// use pseudo rand gen with seed
+mod tests {
+    use std::net::{IpAddr, Ipv4Addr};
+
+    use risingwave_common::util::addr::HostAddr;
+    use risingwave_pb::common::WorkerType;
+    use risingwave_rpc_client::MetaClient;
+    use tokio::time::sleep;
+
+    use super::*;
+
+    async fn setup_n_nodes(n: u16) -> Vec<(JoinHandle<()>, WatchSender<()>)> {
+        let meta_store = Arc::new(MemStore::default());
+
+        let mut node_controllers: Vec<(JoinHandle<()>, WatchSender<()>)> = vec![];
+        for i in 0..n {
+            // TODO: add pseudo random sleep here
+            let node = format!("node{}", i);
+            let err_msg = format!("Meta {} failed in setup", node);
+            let err_msg = err_msg.as_str();
+
+            let info = AddressInfo {
+                addr: node,
```

Comment:
`addr` is used for other components to connect to meta via this address. `listen_addr` is used for listen. We should config `addr` correctly and use it for election. Here in the test you can set them to the same.